src/runtime_src/core/common/api/xrt_kernel.cpp

```diff
@@ -2241,38 +2252,77 @@ class mailbox_impl : public run_impl
     {
       // read arg size bytes from mailbox at arg offset
       // arg size is alwaus a multiple of 4 bytes
-      mbox->mailbox_wait();
+      mbox->mailbox_wait(mailbox_operation::read);
       mbox->kernel->read_register_n(arg.offset(), arg.size() / wsize, data32 + arg.offset() / wsize);
       return run_impl::hs_arg_setter::get_arg_value(arg);
     }
   };
 
   void
-  poll()
+  poll(mailbox_operation mbop)
   {
-    m_ctrlreg = kernel->read_register(0x0);
-    m_busy = m_ctrlreg & (MAILBOX_INPUT_CTRL | MAILBOX_OUTPUT_CTRL);
+    if (mbop == mailbox_operation::write) {
+        uint32_t ctrlreg_write = kernel->read_register(mailbox_input_ctrl_reg);
+        m_busy_write = ctrlreg_write & mailbox_input_ack; //Low - free, High - Busy 
+        return;
+    }
+    
+    if(mbop == mailbox_operation::read) { 
+        uint32_t ctrlreg_read = kernel->read_register(mailbox_output_ctrl_reg);
+        m_busy_read = ctrlreg_read & mailbox_output_ack;//Low - free, High - Busy
+        return;
+    }
   }
 
   void
-  mailbox_idle_or_error()
-  {
-    if (!m_busy)
-      return;
-
-    poll();
+  mailbox_idle_or_error(mailbox_operation mbop)
+  {
+    if (mbop == mailbox_operation::write) { 
+       poll(mailbox_operation::write);
+       if (!m_busy_write)
+           return;
```

Comment:
This return is not required. 
Better to write 
if (m_busy_write) throw an error